joschahenningsen/TUM-Live
# Test course feature for lecturers

### Motivation and Context
Closes #1418.
While this approach is not optimal, as we need to "reserve" special values for test courses, it is relatively simple and builds on top of the existing code in comparison to more complex approaches (e.g., updating the current course model & table to include a `isTestCourse` attribute or adding a completely new test course model).

### Description
- Internally, test courses are handled just like regular courses but belong to a special "test-semester" (where the course's year equals `1234`).
- Updated semester API endpoint and course DAO methods to include an optional `includeTestSemester` flag, which defaults to false. This allows the endpoint & method to return the same information as before (all normal courses), and only if the flag is set, also include the test courses.
- Updated all current occurrences of the semester-fetching (e.g., on the start page, search page, and admin page) accordingly so that the test semesters and courses are not included there. 

### Steps for Testing
Prerequisites:
- 1 Lecturer
- 1 Student

1. Log in as a lecturer
2. Go to the admin panel and press on `+ Create Test Course` to create a new test course
3. The test course should now appear after reloading the page
4. Log out and log in as another user
5. Check that the previously created test course is not visible anywhere and that the semesters don't include the test semester

### Screenshots
How to create a test course:
![image](https://github.com/user-attachments/assets/55e34900-c4df-4150-bf3b-f51d048ff6fd)

How an Admin sees all test courses: 
![image](https://github.com/user-attachments/assets/609d4734-92ba-4a2c-baca-f38aea31431c)
